.github/workflows/pages-deploy.yml.hook

```diff
@@ -1,29 +1,61 @@
-name: 'Automatic build'
+name: "Automatic build and deploy"
 on:
+  workflow_dispatch:
   push:
     branches:
-      - master
+      - main
     paths-ignore:
       - .gitignore
       - README.md
       - LICENSE
 
-jobs:
-  continuous-delivery:
+permissions:
+  contents: read
+  pages: write
+  id-token: write
 
+jobs:
+  build:
+    name: "Build & upload site artifact"
     runs-on: ubuntu-latest
-
+    env:
+      SITE_DIR: "_site"
+      CONFIG_YML: "_config.yml"
+      JEKYLL_ENV: "production"
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
         with:
-          fetch-depth: 0  # for posts's lastmod
+          fetch-depth: 0
+          submodules: true
+
+      - name: Setup Pages
+        id: pages
+        uses: actions/configure-pages@v1
 
       - name: Setup Ruby
         uses: ruby/setup-ruby@v1
         with:
-          ruby-version: 2.7
+          ruby-version: 3.1    # reads from a '.ruby-version' or '.tools-version' file if omitted
           bundler-cache: true
 
-      - name: Deploy
-        run: bash tools/deploy.sh
+      - name: Build site
+        run: bundle exec jekyll build --config "${CONFIG_YML}" --destination "${SITE_DIR}"
+
+      - name: Test site
+        run: bundle exec htmlproofer --disable-external --allow_hash_href "${SITE_DIR}"
+
+      - name: Upload site artifact
+        uses: actions/upload-pages-artifact@v1
+
+  deploy:
+    name: "Deploy site"
+    environment:
+      name: github-pages
+      url: ${{ steps.deployment.outputs.page_url }}
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - name: Deploy to GitHub Pages
+        id: deployment
+        uses: actions/deploy-pages@v1
```

Comment:
These are some of the new actions that work together to deploy the site:
- `actions/configure-pages` performs some setup necessary to enabling the deployment environment.
- `actions/upload-pages-artifact` will grab the content built in the `_site` directory in the runner, create an archive for that content, and upload it. That archive is retained for 1 day within the Action output.
- `actions/deploy-pages` takes the artifact uploaded by `upload-pages-artifact`, and pushes that out to deploy the site.

This workflow follows the [starter workflow for Jekyll](https://github.com/actions/starter-workflows/blob/main/pages/jekyll-gh-pages.yml) closely, but omits the `jekyll-build-pages` action since that is set to Ruby 2.7 (when we can safely use 3.1, in my experience; though I don't use Ruby regularly, so I can't speak to specifics between these versions).

Reply:
This workflow lacks the concurrency limit and also requires additional handling of Jekyll `baseurl` in the _test_. 

Please change it to the following:

```yaml
name: "Automatic build and deploy"
on:
  push:
    branches:
      - master
    paths-ignore:
      - .gitignore
      - README.md
      - LICENSE

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

permissions:
  contents: read
  pages: write
  id-token: write

# Allow one concurrent deployment
concurrency:
  group: "pages"
  cancel-in-progress: true

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          fetch-depth: 0

      - name: Setup Pages
        id: pages
        uses: actions/configure-pages@v1

      - name: Setup Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: 3   # reads from a '.ruby-version' or '.tools-version' file if omitted
          bundler-cache: true

      - name: Build Site
        run: bundle exec jekyll b -d "_site${{ steps.pages.outputs.base_path }}"
        env:
          JEKYLL_ENV: "production"

      - name: Test Site
        run: |
          bundle exec htmlproofer _site --disable-external --check-html --allow_hash_href

      - name: Upload site artifact
        uses: actions/upload-pages-artifact@v1
        with:
          path: "_site${{ steps.pages.outputs.base_path }}"

  deploy:
    name: "Deploy site"
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    runs-on: ubuntu-latest
    needs: build
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v1

```